### ISSUE TITLE
Reach parity with ADBench

### ISSUE THREAD

**Issue by samestep** [gradbench/gradbench#96]
Before we expand to supporting other benchmark programs, we should at least support what is already present in [ADBench](https://github.com/microsoft/ADBench). If I remember correctly, currently we're missing one or two of the ADBench evals, and while I think we support all of the tools, we have not yet migrated all of the evals for all of the tools. This is partially because we were previously planning to handle this "generally" by transpiling from [Adroit](https://adroit-lang.org/), which we have now decided not to do for now after some discussion with @athas.

**Comment by athas** [comment]
Do we actually intend to port all of the ADBench tools? Some are obscure (like the ones for MATLAB) and many of the others are fragile or obsolete, and were never packaged in a reproducible way in the first place.

**Comment by samestep** [comment]
@athas yeah if we decide there are some obscure ones from ADBench that we want to leave out, that's fine. What's the specific list you have in mind?

**Comment by athas** [comment]
ADBench supports the following tools (in the ADBench paper, not sure how many more got added later):

* Adept (C++)
* ADIC2 (C++)
* ADOL-C (C++)
* Ceres (C++)
* clad (C++)
* Tapenade (C)
* DiffSharp (F#)
* ADiGator (MATLAB)
* ADiMat (MATLAB)
* MuPAD (MATLAB)
* ForwardDiff.jl (Julia)
* Autograd (Python)
* Theano (Python)

GradBench already has complete ports of Manual/Finite (which are not AD tools), and complete or partial ports of these tools:

* Tapenade
* DiffSharp
* Autograd

Additionally, GradBench has support for these that are not in ADBench (actually some are in the ADBench repository for sure):

* TensorFlow
* PyTorch
* Futhark
* Zygote
* SciLean
* Jax
* Enzyme

This leaves the following missing:

* Adept (C++): still seems actively maintained, and looks like it always worked well for ADBench. We should also have this.
* ADIC2 (C++): can't find any references newer than 2010, and no code, and the ADBench paper states they could not get it to work.
* ADOL-C (C++): still seems actively maintained, and looks like it always worked well for ADBench. We should also have this.
* Ceres Solver (C++): this is actively maintained, but it's not really an AD library, and it seemed to behave poorly for most of the ADBench benchmarks.
* clad (C++): actively maintained. It did not work properly for the ADBench authors (no support for arrays), but that was many years ago. Might be worth looking into.
* ADiGator (MATLAB): didn't work for the ADBench authors and seems unmaintained.
* ADiMat (MATLAB): can't find any recent references, but it worked well for ADBench. However, I personally have no love for MATLAB, or even cursory interest for that matter. Testing proprietary software in CI is also challenging.
* MuPAD (MATLAB): MATLAB factor again
* ForwardDiff.jl (Julia): Forward-mode is just not going to cut it in general, and we have better AD libraries for Julia now (like Zygote), so no reason for this.
* Theano (Python): even its GitHub repository describes this library in the past tense.

My insightful and completely objective analysis is that we should consider "parity with ADBench" to be the porting of the following:

* Adept (C++)
* ADOL-C (C++)
* clad (C++)

...and of course, finish implementing the evals for those tools that currently only exist in embryonic form.

**Comment by athas** [comment]
[ADBench has a more comprehensive list of their tools here](https://github.com/diku-dk/ADBench/blob/master/STATUS.md#completeness), but it does not invalidate my analysis above.

**Comment by athas** [comment]
While the ADBench paper does talk about clad, it's not part of the ADBench repository (and never was I think), so parity does not require it.

**Comment by athas** [comment]
GradBench has in some ways already exceeded ADBench, and if #397 and #378 is merged will significantly exceed it.

I propose that we define "parity with ADBench" as "adding Zygote implementations", and close this issue after that. That mainly means missing out on a bouquet of flavours of MATLAB AD.

**Comment by athas** [comment]
GradBench now has all of ADBench's Zygote implementations. I suspect they are not *good*, but they are as good as ADBench ever was. That means we have reached parity. Closing this issue.
